### PR TITLE
Make "ignored" classes, interfaces or records ignored rust types as well

### LINKED ISSUE #201: Failure to ignore a class when passed in a container
```toml
[options]
girs_dir = "gir-files"
library = "Gio"
version = "2.0"
min_cfg_version = "2.32"
target_path = "."
work_mode = "normal"
external_libraries = [
    "GLib",
    "GObject",
]

generate = [
    "Gio.AppInfo",
    "Gio.AppLaunchContext",
]
```

```rust
    pub fn get_display<T: IsA<AppInfo>>(&self, info: &T, files: &[File]) -> Option<String> {
        unsafe {
            from_glib_full(ffi::g_app_launch_context_get_display(self.to_glib_none().0, info.to_glib_none().0, files.to_glib_none().0))
        }
    }
```

### PR DESCRIPTION
**Pull request by gkoz**
Fix #201

### CODE REVIEW

File: src/analysis/rust_type.rs [hunk 1]
```diff
@@ -91,14 +112,22 @@ fn rust_type_full(env: &Env, type_id: library::TypeId, nullable: Nullable, ref_m
...
-            rust_type = Err(rust_type_with_prefix.as_str().into());
-        } else {
-            rust_type = rust_type_with_prefix;
+            rust_type = Err(rust_type.as_str().to_owned());
+        }
+        rust_type = rust_type.map_any(|s| format!("{}::{}",
+            crate_name(&env.library.namespace(type_id.ns_id).name), s));
+    }
+
+    if let Err(s) = rust_type {
```
Comment (EPashkin) [inline src/analysis/rust_type.rs]: This condition is true on other errors, like "Unsupported"

File: src/analysis/rust_type.rs [hunk 2]
```diff
@@ -65,9 +65,30 @@ fn rust_type_full(env: &Env, type_id: library::TypeId, nullable: Nullable, ref_m
 
         Enumeration(ref enum_) => Ok(enum_.name.clone()),
         Bitfield(ref bitfield) => Ok(bitfield.name.clone()),
-        Record(ref record) => Ok(record.name.clone()),
-        Interface(ref interface) => Ok(interface.name.clone()),
-        Class(ref klass) => Ok(klass.name.clone()),
+        Record(ref record) => {
+            if env.type_status(&type_id.full_name(&env.library)).ignored() {
```
Comment (EPashkin) [inline src/analysis/rust_type.rs]: Repeated 3 times -> better make function, may include adding /*Ignored*/

File: src/analysis/rust_type.rs (hunk 1 above)
Comment (gkoz) [inline src/analysis/rust_type.rs]: This is getting out of hand, I'll think about adding an enum that would also handle the "Unsupported" and "Ignored" comment.

Comment (EPashkin) [inline src/analysis/rust_type.rs]: Not better just add `check_for_ignored: bool` and check it here?
Then also functions will not needed.

File: src/analysis/rust_type.rs (hunk 2 above)
Comment (gkoz) [inline src/analysis/rust_type.rs]: They can probably be collapsed into a single match arm because using `Type::get_name`.

Comment (EPashkin) [inline src/analysis/rust_type.rs]: Yes, with enumeration and bitfield also

Comment (EPashkin) [inline src/analysis/rust_type.rs]: If you will collapse matches, remove two TODOs about checking `library::Type::get_name()`

File: src/analysis/rust_type.rs [hunk 3]
```diff
@@ -146,29 +193,24 @@ pub fn parameter_rust_type(env: &Env, type_id:library::TypeId,
...
             }
         }
         Alias(ref alias) => {
-            let res = format_parameter(rust_type, direction);
-            if parameter_rust_type(env, alias.typ, direction, nullable, ref_mode).is_ok() {
-                res
-            } else {
-                res.and_then(|s| Err(s))
-            }
+            let rust_type = rust_type.map_any(|s| format_parameter(s, direction));
```
Comment (EPashkin) [inline src/analysis/rust_type.rs]: Can this be moved into `and_then` ?

Comment (EPashkin) [inline src/analysis/rust_type.rs]: Now `Atom` shows as `/*Ignored*/gdk::Atom_` (Ex. in `gtk_text_buffer_register_serialize_tagset`)
Seems this not equal change.

**Comment by gkoz** [comment]
I think I got `Alias` right now.

File: src/analysis/rust_type.rs [hunk 4]
```diff
@@ -146,29 +193,28 @@ pub fn parameter_rust_type(env: &Env, type_id:library::TypeId,
...
         }
         Alias(ref alias) => {
-            let res = format_parameter(rust_type, direction);
-            if parameter_rust_type(env, alias.typ, direction, nullable, ref_mode).is_ok() {
-                res
-            } else {
-                res.and_then(|s| Err(s))
-            }
+            rust_type.and_then(|s| {
+                match parameter_rust_type(env, alias.typ, direction, nullable, ref_mode) {
```
Comment (gkoz) [inline src/analysis/rust_type.rs]: Eh, this should just have been `map_any`

File: src/analysis/rust_type.rs [hunk 5]
```diff
@@ -53,21 +87,36 @@ fn rust_type_full(env: &Env, type_id: library::TypeId, nullable: Nullable, ref_m
...
-                .map_any(|_| alias.name.clone()),
-
-        Enumeration(ref enum_) => Ok(enum_.name.clone()),
-        Bitfield(ref bitfield) => Ok(bitfield.name.clone()),
-        Record(ref record) => Ok(record.name.clone()),
-        Interface(ref interface) => Ok(interface.name.clone()),
-        Class(ref klass) => Ok(klass.name.clone()),
+        Alias(ref alias) => {
+            rust_type_full(env, alias.typ, nullable, ref_mode)
+                .map_any(|_| alias.name.clone())
```
Comment (EPashkin) [inline src/analysis/rust_type.rs]: Now it always aliased name.
Why not `type_.get_name().to_owned()` ?

Comment (gkoz) [inline src/analysis/rust_type.rs]: Just `get_name` without checking if the actual type is not unimplemented?

Comment (EPashkin) [inline src/analysis/rust_type.rs]: Not. In map any instead `alias.name.clone()`

Comment (gkoz) [inline src/analysis/rust_type.rs]: We have the `alias` binding already. I don't see why go through `get_name`. Doesn't it return the same? For consistency?

Comment (EPashkin) [inline src/analysis/rust_type.rs]: For something like `Rectangle` alias for `RectangleInt`. (Just theoretical example)

Comment (EPashkin) [inline src/analysis/rust_type.rs]: Ops. Was wrong. Forget this ;)

**Comment by EPashkin** [comment]
Thanks.

File: src/analysis/rust_type.rs (hunk 5 above)
Comment (gkoz) [inline src/analysis/rust_type.rs]: `type_.get_name` [literally is](`https://github.com/gtk-rs/gir/blob/master/src/library.rs#L384`) the same as `alias.name.clone()`. Whether there's going to be a need to get the name of `alias.typ` is unclear at this point.

Comment (EPashkin) [inline src/analysis/rust_type.rs]: Yes, i mistaken `alias.name` for aliased name.

**Pull request merged**